### PR TITLE
Less pruning on last few frames for IntersectDensePruned

### PR DESCRIPTION
**Pull request by pkufool**
See https://github.com/k2-fsa/icefall/issues/292 and https://github.com/k2-fsa/icefall/issues/292#issuecomment-1093837846 for more details.

### CODE REVIEW

**Comment by videodanchik** [comment]
Alright, after your last changes, then your modifications are the same as mine, but with it, I got doubled, the number of empty utterances in the decoding of multiple datasets. OK let me do decoding with exact your last changes in this PR and return with results tomorrow, for now I can provide some results for decoding with `master` branch and using my custom model for decoding `ami_dev` and `ted_dev` (each column is a separate model, last column is the strongest one in terms of WER, first is the weakest). I present 2 numbers separated by `/` first is empty utterances num., second - the total number of utterances. Beam is 12.
image
this is for beam 15
image

File: k2/csrc/intersect_dense_pruned.cu [hunk 1]
```diff
@@ -570,10 +573,17 @@ class MultiGraphDenseIntersectPruned {
           float best_loglike = max_per_fsa_data[i],
                 dynamic_beam = dynamic_beams_data[i];
           int32_t active_states = arc_end_scores_row_splits1_data[i + 1] -
-                                  arc_end_scores_row_splits1_data[i];
+                                  arc_end_scores_row_splits1_data[i],
+                  final_t = b_fsas_row_splits1[i+1] - b_fsas_row_splits1[i];
```
Comment (danpovey) [inline k2/csrc/intersect_dense_pruned.cu:577]: Is this really final_t.. wouldn't final_t be b_fsas_row_splits1[i+1] - b_fsas_row_splits1[i] - 1, if we intend final_t to be the index of the last frame (i.e. the frame on which final-arcs happen)?

**Comment by videodanchik** [comment]
Ok, once again, with this PR I'm getting the following results (I considered utterance as empty if the reference transcript includes 3 or more words and hypothesis transcript is empty)

Master:
image

This PR:
image
I tried to decode with beam 15, but got CUDA OOM with my batch size, so can provide the table later if necessary, need to wait for finishing the decoding with smaller batch.

Additionally here are the WER I got with master vs this branch.

Master:
image

This PR:
image

File: k2/csrc/intersect_dense_pruned.cu (hunk 1 above)
Comment (videodanchik) [reply k2/csrc/intersect_dense_pruned.cu:577]: I did it the same during my experiments, similar to this line (https://github.com/k2-fsa/k2/blob/510cd726f91d5c6473a37f1714cc13e57b9c056b/k2/csrc/intersect_dense_pruned.cu#L410).

**Comment by danpovey** [comment]
I'm a little confused how this PR could be making WERs worse.  Check if the results particular change toward the end of sentences.  And you could perhaps try printing out the beams, and decoding one minibatch with the vs. the original code (with print statements) to see how the beams change toward the end of the batch.

**Comment by videodanchik** [comment]
I tried to print out the beams and didn't succeed, please provide some instructions on how to do it (may be @pkufool can make a temporal commit to this branch with an example). Instead I provide several examples of lattices where `master` branch code succeeds to produce transcript and this PR fails:

[lattice_1_master.pt.zip](https://github.com/k2-fsa/k2/files/8566906/lattice_1_master.pt.zip)
[lattice_1_this_pr.pt.zip](https://github.com/k2-fsa/k2/files/8566907/lattice_1_this_pr.pt.zip)
[lattice_connected_1_master.pdf](https://github.com/k2-fsa/k2/files/8566909/lattice_connected_1_master.pdf)
[lattice_whole_1_master.pdf](https://github.com/k2-fsa/k2/files/8566910/lattice_whole_1_master.pdf)
[lattice_whole_1_this_pr.pdf](https://github.com/k2-fsa/k2/files/8566911/lattice_whole_1_this_pr.pdf)

[lattice_2_master.pt.zip](https://github.com/k2-fsa/k2/files/8566913/lattice_2_master.pt.zip)
[lattice_2_this_pr.pt.zip](https://github.com/k2-fsa/k2/files/8566914/lattice_2_this_pr.pt.zip)
[lattice_connected_2_master.pdf](https://github.com/k2-fsa/k2/files/8566915/lattice_connected_2_master.pdf)
[lattice_whole_2_master.pdf](https://github.com/k2-fsa/k2/files/8566916/lattice_whole_2_master.pdf)
[lattice_whole_2_this_pr.pdf](https://github.com/k2-fsa/k2/files/8566917/lattice_whole_2_this_pr.pdf)

[lattice_3_master.pt.zip](https://github.com/k2-fsa/k2/files/8566919/lattice_3_master.pt.zip)
[lattice_3_this_pr.pt.zip](https://github.com/k2-fsa/k2/files/8566922/lattice_3_this_pr.pt.zip)
[lattice_connected_3_master.pdf](https://github.com/k2-fsa/k2/files/8566923/lattice_connected_3_master.pdf)
[lattice_whole_3_master.pdf](https://github.com/k2-fsa/k2/files/8566924/lattice_whole_3_master.pdf)
[lattice_whole_3_this_pr.pdf](https://github.com/k2-fsa/k2/files/8566932/lattice_whole_3_this_pr.pdf)

I have an impression that for some reason this PR cuts off the sub-lattice that has a final state. Please, try out yourself on some more or less challenging dataset where you have a `conformer_ctc` pretrained model like GigaSpeech for example. I bet you will find some empty hyps there.

**Comment by videodanchik** [comment]
Ah, okey I have added something like 
```
for (int32_t k = 0; k != num_fsas; ++k) {
    std::cout << dynamic_beams_data[k] << ", ";
}
std::cout << std::endl;
```
and it didn't work for GPU (got segmentation fault), but it worked fine on CPU, I update the dynamic_beam history for a batch from my previous comment.

File: k2/csrc/intersect_dense_pruned.cu (hunk 1 above)
Comment (pkufool) [reply k2/csrc/intersect_dense_pruned.cu:577]: Ah, yes. final_t should be  `b_fsas_row_splits1[i+1] - b_fsas_row_splits1[i] - 1` . But I think this should not be the reason that making the results worse.

**Comment by pkufool** [comment]
> Ah, okey I have added something like
> 
> ```
> for (int32_t k = 0; k != num_fsas; ++k) {
>     std::cout << dynamic_beams_data[k] << ", ";
> }
> std::cout << std::endl;
> ```
> 
> and it didn't work for GPU (got segmentation fault), but it worked fine on CPU, I update the dynamic_beam history for a batch from my previous comment.

You can use `K2_LOG` to print out the whole array.

```
K2_LOG(INFO) << "frame : " << t << "\t beams : " << dynamic_beam_;
```

Note, put this line out of the `K2_EVAL`,  for example, line 612 in this PR, just before we return cutoffs.

**Comment by pkufool** [comment]
I think the reason of WER degradation might be it soon violated the `max-activate` after we increased the `min-activate`.
You can see in the code below, in line 591, the beam value will reset to initial beam then scale with 0.8 if `max-activate` violates. So the beam values might be 12 -> 15 -> 18.75 -> 9.6.

Just a analysis according to the code, let's see the beam values printing out by @videodanchik .

https://github.com/k2-fsa/k2/blob/510cd726f91d5c6473a37f1714cc13e57b9c056b/k2/csrc/intersect_dense_pruned.cu#L568-L598

**Comment by videodanchik** [comment]
Thanks, folks for help! So here are the logs for master and this PR. Each file includes 3 utterances from my previous comment.
Master:
[master.txt](https://github.com/k2-fsa/k2/files/8567870/master.txt)
This PR:
[this_pr.txt](https://github.com/k2-fsa/k2/files/8567873/this_pr.txt)

**Comment by videodanchik** [comment]
And with replacement `b_fsas_row_splits1[i+1] - b_fsas_row_splits1[i] - 1` I I still got first and last utterances empty.
[this_pr_update.txt](https://github.com/k2-fsa/k2/files/8567942/this_pr_update.txt)

**Comment by pkufool** [comment]
That is it, it violates `max-states` at frame `final_t - 3`.
image

image

image

So, do we need to reset the beam  to  the initial value? @danpovey  I think we can just delete that line.

**Comment by danpovey** [comment]
Yes, you could definitly try deletng that line and see what happens.  I hope it doesn't increase the chances of memory blowup in the normal case.

**Comment by videodanchik** [comment]
Hi, I did a couple of experiments and decode several datasets. At first, I want to say that after complete commenting out `if (dynamic_beam > default_beam) dynamic_beam = default_beam;` I got some reduction of WER and the number of empty utterances, though I run into several CUDA OOM errors for beam `15.0`, unless I decreased `max_dur` for batch decoding. Anyway, to add some memory safety I wrapped [`else`](https://github.com/k2-fsa/k2/blob/master/k2/csrc/intersect_dense_pruned.cu#L589-L598) statement like following:

```
} else {
  // We don't want to modify dynamic_beam for the last few frames
  // in order to avoid final states pruning.
  if (t + 5 < final_t) {
    // We violated the max_active constraint -> decrease beam
    if (dynamic_beam > default_beam) dynamic_beam = default_beam;

    // Decrease the beam as long as we have more than
    // max_active active states.
    dynamic_beam *= 0.8;
  }
}
```
to not reset `dynamic_beam` for the last few frames (everything remaining code I keep like in this PR). That helped with OOM, so I tried decoding for `t+5` and `t+10` and got results for my model:
image
As you can see removing pruning the last 5 frames really helped when the beam was 12.0 and further increasing to 10 last frames didn't help. I tried to decode several utterances that still were empty after `one_best_decoding`, though `k2.connect(lattice)` returned not empty lattice this time. I am attaching 2 examples:
[lattice1.pt.zip](https://github.com/k2-fsa/k2/files/9059398/lattice1.pt.zip)
[lattice1.pdf](https://github.com/k2-fsa/k2/files/9059399/lattice1.pdf)
[lattice2.pt.zip](https://github.com/k2-fsa/k2/files/9059401/lattice2.pt.zip)
[lattice2.pdf](https://github.com/k2-fsa/k2/files/9059402/lattice2.pdf)
This is already a very rare cases, but I'm curious if there could be any corner cases bug or something...

**Comment by danpovey** [comment]
@videodanchik  Great, thanks!    Perhaps you could create a branch with your change, modified from the branch in this PR, and we can merge that?

**Comment by videodanchik** [comment]
Sure.

**Comment by videodanchik** [comment]
I make a PR, though there's still some weird stuff is happening with `one_best_decoding` of 2 lattices from my previous comment.

**Comment by pkufool** [comment]
Closed by #1004

**Comment by danpovey** [comment]
I think we overloaded things like
 std::cout << [ragged array]

**Pull request closed without merging**